### PR TITLE
[incubator/mariadb-galera] Add mariadb-galera chart

### PR DESCRIPTION
**Pull request by tongpu**
This charts provides the ability to set up a MariaDB Galera Cluster on Kubernetes leveraging the StatefulSet API.

### CODE REVIEW

**Comment by tongpu** [comment]
I signed it!

**Comment by mgoodness** [comment]
@k8s-bot ok to test

File: incubator/mariadb-galera/LICENSE
Comment (viglesiasce) [inline incubator/mariadb-galera/LICENSE]: Please remove this file. This repository is covered under the Apache license.

File: incubator/mariadb-galera/README.md
Comment (viglesiasce) [inline incubator/mariadb-galera/README.md:1]: Please add a NOTES.txt telling people how to connect. A good example can be found in the mariadb chart:
https://github.com/kubernetes/charts/blob/master/stable/mariadb/templates/NOTES.txt

**Review by viglesiasce — CHANGES_REQUESTED** [review]
Installing this resulted in a crash of the pod with the following error:
2017/05/03 22:43:47 Peer finder exiting
+ '[' '!' -d /var/lib/mysql/mysql ']'
+ echo 'Configure first time mysql'
Configure first time mysql
+ /usr/share/container-scripts/mysql/configure-mysql.sh
+ echo 'Running mysql_install_db ...'
Running mysql_install_db ...
+ mysql_install_db --datadir=/var/lib/mysql
Fatal error Can't create database directory '/var/lib/mysql/mysql'

The latest information about mysql_install_db is available at
https://mariadb.com/kb/en/installing-system-tables-mysql_install_db
mkdir: cannot create directory '/var/lib/mysql/mysql': Permission denied

File: incubator/mariadb-galera/README.md
Comment (ttiurani) [inline incubator/mariadb-galera/README.md]: FYI @tongpu : I fixed the permission problem with this quite ugly hack:

https://github.com/extendedmind/charts/commit/f1455176470bd119d25ecb2cbc9ac4aa7a9ed566

This is related to:

https://github.com/kubernetes/kubernetes/issues/2630

where there is talk that fsGroup thing might also help, but this busybox-initContainer-chown-hack probably is more portable.

**Comment by ttiurani** [comment]
FYI @tongpu : I fixed the permission problem with this quite ugly hack:

extendedmind/charts@f145517

This is related to:

kubernetes/kubernetes#2630

where there is talk that fsGroup thing might also help, but this busybox-initContainer-chown-hack probably is more portable.

**Comment by tongpu** [comment]
@viglesiasce I've updated the charts accordingly.
@ttiurani For the sake of portability (we would like to run the same image in OpenShift and k8s) I'll do it like that for now. According to my discussion with @pmorie `fsGroup` would be the right solution, but I haven't had time to test it yet.

File: incubator/mariadb-galera/README.md
Comment (tongpu) [reply incubator/mariadb-galera/README.md:1]: I've used the one from the mariadb charts as a template, basics really are the same.

**Comment by ttiurani** [comment]
@tongpu thanks for the fix, I now can deploy your PR successfully.

However, I had to make the following patch:

https://github.com/extendedmind/charts/commit/5f6a867a3d75e3c5d718c5b7fa9cee72f967b22a

to add at least one general label to the items created. When all of the labels were generated either from the helm release name, or the chart version, there was no way to target on external "mariadb" service to the statefulset, and make it still possible to upgrade the statefulset without touching the service.

Would you mind just adding also static labels?

Edit: Another thing is that naming the services like this causes big problems for Kuberentes:
I tried to name my release "mariadb-galera-0.1.1" but that created an invalid service name, then I tried to name it "mariadb-galera-011" but that caused Galera connect to fail, as the hostnames were too long. I had to name the release"mg-011" for it to work.

So please also change the generation scheme for the manifest names, to make it possible to set normal names for releases.

**Comment by ttiurani** [comment]
@tongpu also, the persistence.storageClass wasn't actually in use, but this:

https://github.com/extendedmind/charts/commit/0b92ade76a114936db922fe481088dd1a1004c2e

fixed it.

**Comment by viglesiasce** [comment]
@ttiurani is spot on!

Can you update per their requests @tongpu? Also any luck with the CLA?

**Comment by tongpu** [comment]
@viglesiasce I've already signed the CLA before, but my last commit was made with a personal email that was not yet confirmed in my Github account, that triggered the k8s-ci-robot.

@ttiurani I'll make those static labels configurable, otherwise multiple instances of the chart would possibly conflict. @viglesiasce Is `project` as a key acceptable?

Regarding your issues with naming the release `mariadb-galera-0.1.1`. The release name is used in the name of the internal service used for bootstrapping the cluster, so it can't contain any characters illegal for DNS names. I'll extend the template to strip those characters in the service name.

**Comment by tongpu** [comment]
I'm puzzled. Commit 53ef86c failed because I forgot to change the name of the Pods and the StatefulSet. But the current commit (275d848) works for me locally, but the tests fail. I even tried to reproduce the test environment locally by running `helm install --name mariadb-1610 --namespace pr-599-1610 incubator/mariadb-galera` and everything worked as expected. Could this be an issue with the StorageClass? I couldn't find any reference for best practices for that area.

@ttiurani Could you check if this version works for you when you're using `mariadb-galera-0.1.1` as the release name? I still need to add the static labels, but I'm waiting for @viglesiasce feedback on that.

**Comment by tongpu** [comment]
OK I figured out the issues with `persistence.StorageClass` and the tests are now running successful. It would be convenient if the test could wait for 3 Pods to be in state running, because then it would be better suited for the service we're deploying.

**Comment by ttiurani** [comment]
@tongpu Regarding the labels: I can't really understand how having one of many labels just a plain-old-textfield causes problems. I might be missing something, but an extra label doesn't in any way create conflicts, just as long as there are *also* release-specific labels. The generic label only becomes a problem if you target your service on it AND you have multiple galera clusters running.

I most likely have the most common usecase: one Galera cluster running on on my Kube. For me to create my own service to access that cluster really shouldn't force me to configure my own label to the chart, if I could just target one of the labels that doesn't change on version update.

**Comment by tongpu** [comment]
> I most likely have the most common usecase: one Galera cluster running on on my Kube. For me to create my own service to access that cluster really shouldn't force me to configure my own label to the chart, if I could just target one of the labels that doesn't change on version update.

I don't see any issue with that, I just wanted to make sure that I'm following best practices, if there are any for a feature like that.

> I most likely have the most common usecase: one Galera cluster running on on my Kube. For me to create my own service to access that cluster really shouldn't force me to configure my own label to the chart, if I could just target one of the labels that doesn't change on version update.

That will very well be a common usecase, but I also have things like the [OpenServiceBroker API](https://www.openservicebrokerapi.org/) in mind that would, in the future, allow one to run multiple Galera clusters possibly even in the same namespace. I'll implement it for sure, I'm just holding back for the moment so see if the charts maintainers have an opinion on that.

**Comment by ttiurani** [comment]
My problems with the labels were actually caused by my misunderstanding of what a "release" is when using helm: I thought I'd create a new release on upgrade, but that's of course not at all the case. So in that sense, not having a static label at all isn't that big of a problem. I should have named my cluster release based on what data the cluster is used to store, not mariadb-galera-0.1.1.

And I shouldn't have put a version number in my release to begin with, because that really doesn't make a whole lot of sense, given that a helm "release" isn't a release in the way everybody else in the world understands it (eg. "MariaDB release 10.1")  has but instead an *installation*, and helm "release number" is the release.

(Btw. I'm pretty sure I'm not the only one who will make this mistake with confusing Helm terminology. In the Helm glossary it is explicitly even explained that we use the term "release" to mean installation. Why not just then use "installation" as that isn't even a reserved term yet? :) )

**Comment by viglesiasce** [comment]
Sorry for the delay here but I am not able to get this to work with Helm 2.4.2 and K8s 1.6.4. 

I am not able to login as root or a created user with either the randomly created password or one set explicitly. 

```
viglesias@dev-f340:~/charts/incubator$ helm version
Client: &version.Version{SemVer:"v2.4.2", GitCommit:"82d8e9498d96535cc6787a6a9194a76161d29b4c", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.4.2", GitCommit:"82d8e9498d96535cc6787a6a9194a76161d29b4c", GitTreeState:"clean"}
viglesias@dev-f340:~/charts/incubator$ kubectl version
Client Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.4", GitCommit:"d6f433224538d4f9ca2f7ae19b252e6fcb66a3ae", GitTreeState:"clean", BuildDate:"2017-05-19T18:44:27Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.4", GitCommit:"d6f433224538d4f9ca2f7ae19b252e6fcb66a3ae", GitTreeState:"clean", BuildDate:"2017-05-19T18:33:17Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
viglesias@dev-f340:~/charts/incubator$
```

**Comment by tongpu** [comment]
> I am not able to login as root or a created user with either the randomly created password or one set explicitly.

@viglesiasce OK I see the issue. In the Chart we're creating a dedicated service for bootstrapping which we assign the IP None, because of that the commands I provided won't work. Should we simply remove the `clusterIP: None` in the Service template? Are there any best practices related to StatefulSet yet?

**Comment by unguiculus** [comment]
@tongpu Statefulsets require a headless service. So, this should not be changed.

File: incubator/mariadb-galera/.helmignore
Comment (chenzhiwei) [inline incubator/mariadb-galera/.helmignore:10]: why adding too many `.git` here?

File: incubator/mariadb-galera/values.yaml
Comment (chenzhiwei) [inline incubator/mariadb-galera/values.yaml:5]: Could you use a meaningful tag? It's better to tell user the version of mariadb in the tag.

**Review by chenzhiwei — COMMENTED** [review]
see my inline comments.

**Comment by unguiculus** [comment]
Marking this as stale. Please update within one week.

**Comment by unguiculus** [comment]
Closing this as stale. Feel free to reopen when/if you decide to work on it again.

**Pull request closed without merging**
